File: packages/bdb/src/tests/merkle-tree-ref.test.ts
import { MerkleRef } from '../lib/merkle-tree-ref';
import { createTree, type IMerkleTree, type HashedItem } from 'merkle-tree';
import { RandomUuidGenerator } from 'utils';

const uuidGenerator = new RandomUuidGenerator();

//
// Creates a simple HashedItem for testing.
//
function makeItem(name: string): HashedItem {
    return {
        name,
        hash: Buffer.from(name),
        length: name.length,
        lastModified: new Date('2024-01-01'),
    };
}

//
// Creates a new empty tree for use as a stub.
//
function makeTree(): IMerkleTree<undefined> {
    return createTree<undefined>(uuidGenerator.generate());
}

//
// Builds a MerkleRef backed by in-memory state (no real storage).
//
function makeRef(initialTree: IMerkleTree<undefined> | undefined = undefined) {
    let stored: IMerkleTree<undefined> | undefined = initialTree;
    let saveCount = 0;
    let deleteCount = 0;

    const loader = async () => stored;
    const saver = async (tree: IMerkleTree<undefined>) => {
        stored = tree;
        saveCount++;
    };
    const deleter = async () => {
        stored = undefined;
        deleteCount++;
    };
    const creator = async () => makeTree();

    const ref = new MerkleRef(loader, saver, deleter, creator);

    return { ref, getSaved: () => stored, getSaveCount: () => saveCount, getDeleteCount: () => deleteCount };
}

test('get returns undefined when loader returns undefined', async () => {
    const { ref } = makeRef(undefined);
    const tree = await ref.get();
    expect(tree).toBeUndefined();
});

test('get returns the tree when loader returns one', async () => {
    const tree = makeTree();
    const { ref } = makeRef(tree);
    const result = await ref.get();
    expect(result).toBe(tree);
});

test('get only calls loader once (caches result)', async () => {
    let callCount = 0;
    const tree = makeTree();
    const ref = new MerkleRef(
        async () => { callCount++; return tree; },
        async () => {},
        async () => {},
        async () => makeTree(),
    );

    await ref.get();
    await ref.get();
    expect(callCount).toBe(1);
});

test('upsert creates a tree via creator when tree is undefined', async () => {
    const { ref } = makeRef(undefined);
    await ref.upsert(makeItem('file1'));
    const tree = await ref.get();
    expect(tree).toBeDefined();
    expect(tree!.sort).toBeDefined();
});

test('upsert inserts item into existing tree', async () => {
    const { ref } = makeRef(makeTree());
    await ref.upsert(makeItem('file1'));
    const tree = await ref.get();
    expect(tree!.sort).toBeDefined();
    expect(tree!.sort!.leafCount).toBe(1);
});

test('upsert marks the ref as dirty', async () => {
    const { ref, getSaveCount } = makeRef(makeTree());
    await ref.upsert(makeItem('file1'));
    await ref.commit();
    expect(getSaveCount()).toBe(1);
});

test('remove on empty tree is a no-op', async () => {
    const { ref } = makeRef(undefined);
    await expect(ref.remove('file1')).resolves.toBeUndefined();
    const tree = await ref.get();
    expect(tree).toBeUndefined();
});

test('remove deletes an existing item', async () => {
    const { ref } = makeRef(makeTree());
    await ref.upsert(makeItem('file1'));
    await ref.commit();

    // flush to reset dirty, then remove
    ref.flush();
    await ref.remove('file1');
    const tree = await ref.get();
    // After removing the only item, _tree is set to undefined internally
    expect(tree).toBeUndefined();
});

test('remove sets tree to undefined when it becomes empty', async () => {
    const { ref } = makeRef(makeTree());
    await ref.upsert(makeItem('file1'));

    // commit and flush so we can remove cleanly
    await ref.commit();
    ref.flush();

    await ref.remove('file1');
    await ref.commit();
    const tree = await ref.get();
    // After commit/flush cycle, stored tree is cleared by deleter
    // get() now reloads from storage, which was deleted
    expect(tree).toBeUndefined();
});

test('commit saves the tree when dirty', async () => {
    const { ref, getSaveCount } = makeRef(makeTree());
    await ref.upsert(makeItem('file1'));
    await ref.commit();
    expect(getSaveCount()).toBe(1);
});

test('commit calls deleter when tree is empty after remove', async () => {
    const { ref, getDeleteCount } = makeRef(makeTree());
    await ref.upsert(makeItem('file1'));
    await ref.commit();

    ref.flush();
    await ref.remove('file1');
    await ref.commit();
    expect(getDeleteCount()).toBe(1);
});

test('commit is a no-op when not dirty', async () => {
    const { ref, getSaveCount, getDeleteCount } = makeRef(makeTree());
    await ref.commit();
    expect(getSaveCount()).toBe(0);
    expect(getDeleteCount()).toBe(0);
});

test('commit clears the dirty flag', async () => {
    const { ref, getSaveCount } = makeRef(makeTree());
    await ref.upsert(makeItem('file1'));
    await ref.commit();
    // second commit should be a no-op
    await ref.commit();
    expect(getSaveCount()).toBe(1);
});

test('flush resets loaded state so next get reloads from storage', async () => {
    let callCount = 0;
    const tree = makeTree();
    const ref = new MerkleRef(
        async () => { callCount++; return tree; },
        async () => {},
        async () => {},
        async () => makeTree(),
    );

    await ref.get();
    ref.flush();
    await ref.get();
    expect(callCount).toBe(2);
});

test('flush throws when dirty', async () => {
    const { ref } = makeRef(makeTree());
    await ref.upsert(makeItem('file1'));
    expect(() => ref.flush()).toThrow();
});

test('flush does not throw when not dirty', async () => {
    const { ref } = makeRef(makeTree());
    expect(() => ref.flush()).not.toThrow();
});

test('multiple upserts accumulate items in the tree', async () => {
    const { ref } = makeRef(makeTree());
    await ref.upsert(makeItem('file1'));
    await ref.upsert(makeItem('file2'));
    await ref.upsert(makeItem('file3'));
    const tree = await ref.get();
    expect(tree!.sort).toBeDefined();
    expect(tree!.sort!.leafCount).toBe(3);
});

File: packages/bdb-zig/src/test/merkle-tree-ref.test.zig
const std = @import("std");
const bdb = @import("bdb-zig");
const utils = @import("utils-zig");
const merkle_tree_zig = @import("merkle-tree-zig");
const merkle_tree = merkle_tree_zig.merkle_tree;
const IMerkleTree = merkle_tree.IMerkleTree;
const HashedItem = merkle_tree.HashedItem;
const MerkleRef = bdb.merkle_tree_ref.MerkleRef;

const io = std.testing.io;

//
// Creates a simple HashedItem for testing.
//
fn makeItem(name: []const u8) HashedItem {
    return .{
        .name = name,
        .hash = name,
        .length = name.len,
        .lastModified = 1704067200000,
    };
}

//
// Generates ids for the trees the tests create.
//
var uuid_generator: utils.test_uuid_generator.TestUuidGenerator = .{};

//
// Creates a new empty tree for use as a stub.
//
fn makeTree(allocator: std.mem.Allocator) !IMerkleTree {
    return merkle_tree.createTree(try uuid_generator.generate(allocator));
}

//
// In-memory state behind a MerkleRef (TypeScript: the variables captured by makeRef's arrow functions).
//
const RefState = struct {
    // The stored tree (null when there is none).
    stored: ?IMerkleTree,

    // Number of saver calls.
    saveCount: u32 = 0,

    // Number of deleter calls.
    deleteCount: u32 = 0,

    // Number of loader calls.
    loadCount: u32 = 0,

    //
    // Returns the stored tree.
    //
    fn loader(context: *anyopaque, allocator: std.mem.Allocator, ioValue: std.Io) anyerror!?IMerkleTree {
        _ = allocator;
        _ = ioValue;
        const self: *RefState = @ptrCast(@alignCast(context));
        self.loadCount += 1;
        return self.stored;
    }

    //
    // Stores the tree.
    //
    fn saver(context: *anyopaque, allocator: std.mem.Allocator, ioValue: std.Io, tree: *IMerkleTree) anyerror!void {
        _ = allocator;
        _ = ioValue;
        const self: *RefState = @ptrCast(@alignCast(context));
        self.stored = tree.*;
        self.saveCount += 1;
    }

    //
    // Deletes the stored tree.
    //
    fn deleter(context: *anyopaque, allocator: std.mem.Allocator, ioValue: std.Io) anyerror!void {
        _ = allocator;
        _ = ioValue;
        const self: *RefState = @ptrCast(@alignCast(context));
        self.stored = null;
        self.deleteCount += 1;
    }

    //
    // Creates a new empty tree.
    //
    fn creator(context: *anyopaque, allocator: std.mem.Allocator, ioValue: std.Io) anyerror!IMerkleTree {
        _ = context;
        _ = ioValue;
        return makeTree(allocator);
    }

    //
    // Builds a MerkleRef backed by this state (TypeScript: makeRef).
    //
    fn makeRef(self: *RefState, allocator: std.mem.Allocator) MerkleRef {
        return MerkleRef.init(allocator, self, loader, saver, deleter, creator);
    }
};

test "get returns undefined when loader returns undefined" {
    var arena = std.heap.ArenaAllocator.init(std.testing.allocator);
    defer arena.deinit();
    var state: RefState = .{ .stored = null };
    var ref = state.makeRef(arena.allocator());
    try std.testing.expectEqual(@as(?*IMerkleTree, null), try ref.get(io));
}

test "get returns the tree when loader returns one" {
    var arena = std.heap.ArenaAllocator.init(std.testing.allocator);
    defer arena.deinit();
    const tree = try makeTree(arena.allocator());
    var state: RefState = .{ .stored = tree };
    var ref = state.makeRef(arena.allocator());
    const result = (try ref.get(io)).?;
    try std.testing.expectEqualStrings(tree.id, result.id);
}

test "get only calls loader once (caches result)" {
    var arena = std.heap.ArenaAllocator.init(std.testing.allocator);
    defer arena.deinit();
    var state: RefState = .{ .stored = try makeTree(arena.allocator()) };
    var ref = state.makeRef(arena.allocator());
    _ = try ref.get(io);
    _ = try ref.get(io);
    try std.testing.expectEqual(@as(u32, 1), state.loadCount);
}

test "upsert creates a tree via creator when tree is undefined" {
    var arena = std.heap.ArenaAllocator.init(std.testing.allocator);
    defer arena.deinit();
    var state: RefState = .{ .stored = null };
    var ref = state.makeRef(arena.allocator());
    try ref.upsert(io, makeItem("file1"));
    const tree = (try ref.get(io)).?;
    try std.testing.expect(tree.sort != null);
}

test "upsert inserts item into existing tree" {
    var arena = std.heap.ArenaAllocator.init(std.testing.allocator);
    defer arena.deinit();
    var state: RefState = .{ .stored = try makeTree(arena.allocator()) };
    var ref = state.makeRef(arena.allocator());
    try ref.upsert(io, makeItem("file1"));
    const tree = (try ref.get(io)).?;
    try std.testing.expectEqual(@as(u32, 1), tree.sort.?.leafCount);
}

test "upsert marks the ref as dirty" {
    var arena = std.heap.ArenaAllocator.init(std.testing.allocator);
    defer arena.deinit();
    var state: RefState = .{ .stored = try makeTree(arena.allocator()) };
    var ref = state.makeRef(arena.allocator());
    try ref.upsert(io, makeItem("file1"));
    try ref.commit(io);
    try std.testing.expectEqual(@as(u32, 1), state.saveCount);
}

test "remove on empty tree is a no-op" {
    var arena = std.heap.ArenaAllocator.init(std.testing.allocator);
    defer arena.deinit();
    var state: RefState = .{ .stored = null };
    var ref = state.makeRef(arena.allocator());
    try ref.remove(io, "file1");
    try std.testing.expectEqual(@as(?*IMerkleTree, null), try ref.get(io));
}

test "remove deletes an existing item" {
    var arena = std.heap.ArenaAllocator.init(std.testing.allocator);
    defer arena.deinit();
    var state: RefState = .{ .stored = try makeTree(arena.allocator()) };
    var ref = state.makeRef(arena.allocator());
    try ref.upsert(io, makeItem("file1"));
    try ref.commit(io);

    // (Zig: flush is not ported; the removal works on the cached tree.)
    try ref.remove(io, "file1");
    // After removing the only item, _tree is set to undefined internally
    try std.testing.expectEqual(@as(?*IMerkleTree, null), try ref.get(io));
}

test "remove sets tree to undefined when it becomes empty" {
    var arena = std.heap.ArenaAllocator.init(std.testing.allocator);
    defer arena.deinit();
    var state: RefState = .{ .stored = try makeTree(arena.allocator()) };
    var ref = state.makeRef(arena.allocator());
    try ref.upsert(io, makeItem("file1"));
    try ref.commit(io);

    try ref.remove(io, "file1");
    try ref.commit(io);
    try std.testing.expectEqual(@as(?*IMerkleTree, null), try ref.get(io));
    try std.testing.expectEqual(@as(?IMerkleTree, null), state.stored);
}

test "commit saves the tree when dirty" {
    var arena = std.heap.ArenaAllocator.init(std.testing.allocator);
    defer arena.deinit();
    var state: RefState = .{ .stored = try makeTree(arena.allocator()) };
    var ref = state.makeRef(arena.allocator());
    try ref.upsert(io, makeItem("file1"));
    try ref.commit(io);
    try std.testing.expectEqual(@as(u32, 1), state.saveCount);
}

test "commit calls deleter when tree is empty after remove" {
    var arena = std.heap.ArenaAllocator.init(std.testing.allocator);
    defer arena.deinit();
    var state: RefState = .{ .stored = try makeTree(arena.allocator()) };
    var ref = state.makeRef(arena.allocator());
    try ref.upsert(io, makeItem("file1"));
    try ref.commit(io);

    try ref.remove(io, "file1");
    try ref.commit(io);
    try std.testing.expectEqual(@as(u32, 1), state.deleteCount);
}

test "commit is a no-op when not dirty" {
    var arena = std.heap.ArenaAllocator.init(std.testing.allocator);
    defer arena.deinit();
    var state: RefState = .{ .stored = try makeTree(arena.allocator()) };
    var ref = state.makeRef(arena.allocator());
    try ref.commit(io);
    try std.testing.expectEqual(@as(u32, 0), state.saveCount);
    try std.testing.expectEqual(@as(u32, 0), state.deleteCount);
}

test "commit clears the dirty flag" {
    var arena = std.heap.ArenaAllocator.init(std.testing.allocator);
    defer arena.deinit();
    var state: RefState = .{ .stored = try makeTree(arena.allocator()) };
    var ref = state.makeRef(arena.allocator());
    try ref.upsert(io, makeItem("file1"));
    try ref.commit(io);
    // second commit should be a no-op
    try ref.commit(io);
    try std.testing.expectEqual(@as(u32, 1), state.saveCount);
}

test "multiple upserts accumulate items in the tree" {
    var arena = std.heap.ArenaAllocator.init(std.testing.allocator);
    defer arena.deinit();
    var state: RefState = .{ .stored = try makeTree(arena.allocator()) };
    var ref = state.makeRef(arena.allocator());
    try ref.upsert(io, makeItem("file1"));
    try ref.upsert(io, makeItem("file2"));
    try ref.upsert(io, makeItem("file3"));
    const tree = (try ref.get(io)).?;
    try std.testing.expectEqual(@as(u32, 3), tree.sort.?.leafCount);
}

test "get rebuilds the merkle tree when the tree is dirty" {
    var arena = std.heap.ArenaAllocator.init(std.testing.allocator);
    defer arena.deinit();
    var state: RefState = .{ .stored = try makeTree(arena.allocator()) };
    var ref = state.makeRef(arena.allocator());
    try ref.upsert(io, makeItem("file1"));
    const tree = (try ref.get(io)).?;
    try std.testing.expect(!tree.dirty);
    try std.testing.expect(tree.merkle != null);
}
